githubapp/githubapp.go

```diff
@@ -0,0 +1,229 @@
+		f.accessTokenURLPattern = pattern
+	}
+}
+
+// WithJWTTokenExpiration is an option that allows overriding the default expiration
+// date of the application JWTs.
+func WithJWTTokenExpiration(exp time.Duration) GitHubAppConfigOption {
+	return func(f *GitHubAppConfig) {
+		f.jwtTokenExpiration = exp
+	}
+}
+
+// WithJWTTokenCaching is an option that tells the GitHub app to cache its
+// JWT App tokens. The amount of time that the tokens are cached is based
+// on the provided `beforeExp` parameter + the configured token expiration.
+// This results in a cache expiration of <token expiration> - <beforeExp>.
+func WithJWTTokenCaching(beforeExp time.Duration) GitHubAppConfigOption {
+	return func(f *GitHubAppConfig) {
+		exp := f.jwtTokenExpiration
+		f.jwtCacheDuration = exp - beforeExp
+	}
+}
+
+// NewConfig creates a new configuration object containing the three primary required
+// configuration values. Options allow for the customization of rarely used configuration
+// values. Options are evaluated in order from first to last.
+func NewConfig(appID, installationID string, privateKey *rsa.PrivateKey, opts ...GitHubAppConfigOption) *GitHubAppConfig {
+	config := GitHubAppConfig{
+		AppID:                 appID,
+		InstallationID:        installationID,
+		PrivateKey:            privateKey,
+		accessTokenURLPattern: defaultGitHubAccessTokenURLPattern,
+		jwtTokenExpiration:    9 * time.Minute,
+		jwtCacheDuration:      0 * time.Nanosecond,
+	}
+	for _, opt := range opts {
+		opt(&config)
+	}
+	return &config
+}
+
+// TokenRequest is a struct that contains the list of repositories and the
+// requested permissions / scopes that are requested when generating a
+// new installation access token.
+type TokenRequest struct {
+	Repositories []string          `json:"repositories"`
+	Permissions  map[string]string `json:"permissions"`
+}
+
+// GitHubApp is an object that can be used to generate application level JWTs
+// or to request an OIDC token on behalf of an installation.
+type GitHubApp struct {
+	config   *GitHubAppConfig
+	jwtCache *cache.Cache[[]byte]
+}
+
+// New creates a GitHubApp instance based on the provided
+// GitHubAppConfig object.
+func New(config *GitHubAppConfig) *GitHubApp {
+	app := GitHubApp{
+		config: config,
+	}
+	if config.jwtCacheDuration != 0 {
+		app.jwtCache = cache.New[[]byte](config.jwtCacheDuration)
+	}
+	return &app
+}
+
+// AppToken creates a signed JWT to authenticate a GitHub app
+// so that it can make API calls to GitHub.
+func (g *GitHubApp) AppToken() ([]byte, error) {
+	// If token caching is enabled, look first in the cache
+	if g.jwtCache != nil {
+		// Check for a valid JWT in the cache
+		signedJwt, ok := g.jwtCache.Lookup(cacheKey)
+		if !ok {
+			// Create a JWT for reading instance information from GitHub
+			signedJwt, err := g.generateAppJWT()
+			if err != nil {
+				return nil, fmt.Errorf("error generating the JWT for GitHub app access: %w", err)
+			}
+			g.jwtCache.Set(cacheKey, signedJwt)
```

Comment:
nit: should we set a variable and do this logic once? i.e. `shouldFetch` or something like that

Reply:
ack